rfcs/2023-01-06-layout.md

```diff
@@ -0,0 +1,461 @@
+| Grid **(web only)** | `grid*`                                     | [List of props](https://github.com/styled-system/styled-system/blob/master/docs/table.md#grid-layout)\* referenced from styled-system. |
+
+> **Note**
+>
+> Not all CSS grid properties are covered in `styled-system` spec. These should be referenced from MDN during implementation. Here's a nice reference from [CSS Tricks](https://css-tricks.com/snippets/css/complete-guide-grid).
+
+**Spacing values**
+
+- All spacing based properties should accept spacing tokens i.e. `spacing.1`, `spacing.2`, etc.
+- TypeScript auto suggestion hints should also show the `px` value associated with a spacing token, this helps in translating design from Figma to code since designs would have these in `px`.
+- They should also accept `number` or `string` values eg. `1`, `'1px'`, `'100%'`, `'auto'`, `'1fr'` etc. Numeric literals imply the pixel value (as in vanilla React), if someone wants to use a spacing token it must be prefixed like `spacing.*`.
+- **Tradeoff:** The ability to pass any values here can lead to inconsistencies, where users may be passing values not in tokens (eg. `7px`) or passing the token value directly rather than using the actual token (eg. `2px` instead of `spacing.1`). However, this flexibility is needed in order to ease up building UIs. There are genuine usecases of using percentages and spacings not in spacing tokens (especially in landing pages). Also, spacing tokens are capped (`spacing.11`) so it wouldn't be feasible for users to specify larger spacings (without updating spacing tokens in theme), in such cases they would've to rely on custom components or styling.
+
+**Negative values**
+
+There are some usecases where you may want to pass negative values (eg. negative margin values to create full-bleed layouts that break out of main content). In order to support this we would need spacing values to support negative tokens as well. Instead of negative spacing tokens like `-spacing.1`, `-spacing.2`, etc. we'll let users pass negative literals or pixel values directly eg. `-4`, `-4px`, etc.
+
+**Few points to note for React (web)**
+
+- We should avoid supporting shorthands to save from confusions _(should I use `mb` or `marginBottom`)_.
+- Flex gap has ~91% global [coverage](https://caniuse.com/?search=flex%20gap) (support added in some browsers mid 2020). Although the support looks decent, this might be problematic for some products supporting these browsers. There is a [post CSS plugin](https://github.com/gavinmcfarland/flex-gap-polyfill) which can polyfill this _(with some known limitations)_. Alternative is to internally translate gap properties to utilize CSS selectors such as [lobotomized owl](https://alistapart.com/article/axiomatic-css-and-lobotomized-owls/) `* + *`.
```

Comment:
Let's let the consumers take responsibility for the polyfill. If they're using flex gap (or any other property for that matter), they should be aware of its browser compatibility and use it accordingly. We should worry about this property only if we're also using it internally for a blade component since that would reduce it's coverage and impact our consumers using the component.

Reply:
The polyfill has some known limitation plus integration efforts for post css, so it won't work for all cases. Most foolproof alternative I find is internally handling the `gap` if we've to with `* + *` margins but its spacing algorithm would be different from the one provided by flexbox.

Chakra handles it in a way where they have a `Flex` component which uses the native `gap` but their `Stack` component uses `* + *`. Along similar lines I feel if we're providing a `Flex` component it should use the native gap functionality. If a product has legit usecases for supporting older browsers without compromising on the UI they can use margin properties on siblings.

Reply:
Updated in https://github.com/razorpay/blade/pull/950/commits/f4e3b6cebb79e8d1a3b72ed54eb53ed67c55dec4 to clarify this will be a native `gap` property with an alternative to rely on margin properties otherwise